lib/core/src/Cardano/Wallet/Network.hs

```diff
@@ -140,3 +148,66 @@ defaultRetryPolicy =
     limitRetriesByCumulativeDelay (60 * second) (exponentialBackoff 10000)
   where
     second = 1000*1000
+
+type RestoreHandle block = (NE.NonEmpty block -> BlockHeader -> IO Bool)
+
+class Cursor s where
+    toHeader :: s -> BlockHeader
+
+restore
+    :: Cursor s
+    => NetworkLayer IO tx block
+    -> Trace IO Text
+    -> IO s
+    -> RestoreHandle block
+    -> (block -> BlockHeader)
+    -> IO ()
+restore nl tr get f blockTip = do
+    s <- get
+    runExceptT (networkTip nl) >>= \case
+        Right nwTip -> do
+            restoreStep (toHeader s, nwTip)
+        Left e -> do
+            logError tr $ "Failed to get network tip: " +|| e ||+ ""
+            restoreSleep (toHeader s)
+  where
+    restoreStep
+        :: (BlockHeader, BlockHeader)
+        -> IO ()
+    restoreStep (localTip, nodeTip) = do
+        runExceptT (nextBlocks nl localTip) >>= \case
+            Left e -> do
+                logError tr $ "Failed to get next blocks: " +|| e ||+ "."
+                restoreSleep localTip
+            Right [] -> do
+                logDebug tr "Wallet restored."
+                restoreSleep localTip
+            Right (blockFirst : blocksRest) -> do
+                let blocks = blockFirst NE.:| blocksRest
+
+                -- Get the tip of the last block, /before/ advancing the state
+                -- to that point. We need a separate accessor for this, but it
+                -- might be useful for performance.
+                let nextLocalTip = blockTip $ NE.last $ blocks
+                f blocks nodeTip >>= \case
```

Comment:
I suggest a more expressive name than `f` here :sweat_smile: ... Short name are good in short function or, when we don't make assumption on variables. Here, that `f` is a `(NE.NonEmpty block -> BlockHeader -> IO Bool)`, which says quite a lot already on what `f` could be.